src/tools/relooper.jl

```diff
@@ -0,0 +1,129 @@
+    nextargs = []
+    for i in succ
+        br = findfirst(x->x.block==i, brs)
+        push!(nextargs,
+              if br === nothing
+                  []
+              else
+                  [
+                   arg isa Variable ? get!(gensym, symbols, arg) : arg
+                   for arg in brs[br].args
+                  ]
+              end
+             )
+    end
+    return nextargs
+end
+
+
+# NOTE: this will break if arbitrary gotos are used.
+function restructure(
+                     blocks::Vector,
+                     args,
+                     symbols = Dict(),
+                     entries = 1,
+                     loops = [],
+                     cond = nothing
+                    )
+
+    if isempty(entries)
+        return []
+    elseif length(entries) == 1
+        l = entries[1]
+        block = blocks[l]
+
+        assign_args = Expr.(:(=), get!.(gensym, Ref(symbols), arguments(block)), args[1])
+
+        if !isempty(loops)
+            l == loops[end] && return [assign_args..., :(continue)]
+            l in loops && return [assign_args..., :(break)]
+            l == successors(blocks[loops[end]])[end] && return [assign_args..., :(break)]
+        end
+
+        code = extract_expr(block, symbols)
+        cond = condition(block)
+
+        if isreturn(block)
+            val = returnvalue(block)
+            if val isa Variable
+                val = get!(gensym, symbols, val)
+            end
+            code = [code..., cond === nothing ?
+                    :(return $val) : Expr(:if, get!(gensym, symbols, cond), :(return $val))]
+        end
+
+        succ = [i.id for i in successors(block)]
+        isempty(succ) && return [assign_args..., code...]
+        brs = branches(block)
+        nargs = nextargs(succ, branches(block), symbols)
+
+        if !accessible(block, loops)
+            return [
+                    assign_args...,
+                    code...,
+                    restructure(blocks, nargs, symbols, succ, loops, cond)...
+                   ]
+        else
+            return [
+                    assign_args...,
+                    Expr(:while, true,
+                         Expr(:block,
+                              code...,
+                              restructure(blocks, nargs, symbols, succ, [loops...,l], cond)...
+                             )
+                        )
+                   ]
+        end
+    elseif length(entries) == 2 && cond != nothing
+        return [
+                Expr(:if, get!(gensym, symbols, cond),
+                     Expr(:block, restructure(blocks, [args[2]], symbols, entries[2], loops)...),
+                     Expr(:block, restructure(blocks, [args[1]], symbols, entries[1], loops)...))
+               ]
+    else
+        @show entries
+        error("Rebuild error")
+    end
+end
+
+function restructure(code::IR, args=[])
+    return Expr(:block, restructure(blocks(code), [["foo", args...]])...)
```

Comment:
@MikeInnes Do you have any test cases that actually use the first argument passed to the entry point? It says in the docs that this should be the function itself and is used for some closures and callable objects, but I wasn't able to produce an example.

Reply:
You can define a closure like `f = let y = 3; x -> x*y; end` or define a struct and make it callable.